src/netlink_nlctrl.c

```diff
@@ -0,0 +1,259 @@
+/*
+ * Copyright (c) 2017-2024 The strace developers.
+ * All rights reserved.
+ *
+ * SPDX-License-Identifier: LGPL-2.1-or-later
+ */
+
+#include "defs.h"
+#include "netlink_generic.h"
+#include "nlattr.h"
+#include <linux/cgroupstats.h>
+#include <linux/devlink.h>
+#include <linux/ethtool_netlink.h>
+#include <linux/ioam6_genl.h>
+#include <linux/mptcp_pm.h>
+#include <linux/netdev.h>
+#include <linux/nl80211.h>
+#include <linux/seg6_genl.h>
+#include <linux/tcp_metrics.h>
+#include <linux/thermal.h>
+
+#include "xlat/genl_ctrl_cmd.h"
+#include "xlat/genl_ctrl_attr.h"
+#include "xlat/genl_ctrl_attr_op.h"
+#include "xlat/genl_ctrl_attr_op_flags.h"
+#include "xlat/genl_ctrl_attr_mcast_grp.h"
+#include "xlat/genl_ctrl_attr_policy.h"
+#include "xlat/genl_devlink_cmd.h"
+#include "xlat/genl_ethtool_msg_send.h"
+#include "xlat/genl_ethtool_msg_recv.h"
+#include "xlat/genl_ioam6_cmd.h"
+#include "xlat/genl_mptcp_pm_cmd.h"
+#include "xlat/genl_netdev_cmd.h"
+#include "xlat/genl_nl80211_cmd.h"
+#include "xlat/genl_seg6_cmd.h"
+#include "xlat/genl_taskstats_cmd.h"
+#include "xlat/genl_tcp_metrics_cmd.h"
+#include "xlat/genl_thermal_cmd.h"
+#include "xlat/nl_attr_type.h"
+#include "xlat/nl_policy_type_attr.h"
+
+#define ARG_PAIR(arg) { arg, arg }
+static const struct {
+	const char *family;
+	const struct xlat *cmd[2];
+} family_names[] = {
+	{ "nlctrl", ARG_PAIR(genl_ctrl_cmd) },
+	{ DEVLINK_GENL_NAME, ARG_PAIR(genl_devlink_cmd) },
+	{ ETHTOOL_GENL_NAME, { genl_ethtool_msg_send, genl_ethtool_msg_recv } },
+	{ IOAM6_GENL_NAME, ARG_PAIR(genl_ioam6_cmd) },
+	{ MPTCP_PM_NAME, ARG_PAIR(genl_mptcp_pm_cmd) },
+	{ NETDEV_FAMILY_NAME, ARG_PAIR(genl_netdev_cmd) },
+	{ NL80211_GENL_NAME, ARG_PAIR(genl_nl80211_cmd) },
+	{ SEG6_GENL_NAME, ARG_PAIR(genl_seg6_cmd) },
+	{ TASKSTATS_GENL_NAME, ARG_PAIR(genl_taskstats_cmd) },
+	{ TCP_METRICS_GENL_NAME, ARG_PAIR(genl_tcp_metrics_cmd) },
+	{ THERMAL_GENL_FAMILY_NAME, ARG_PAIR(genl_thermal_cmd) },
+};
+#undef ARG_PAIR
+
+static
+DECL_NLA(ctrl_attr_family_name)
+{
+	char name[GENL_NAMSIZ];
+
+	if (len <= sizeof(name) && umoven(tcp, addr, len, name) == 0) {
+		unsigned int *idx = (void *) opaque_data;
+		for (unsigned int i = 0; i < ARRAY_SIZE(family_names); ++i) {
+			if (strncmp(name, family_names[i].family, len) == 0) {
+				*idx = i;
+				break;
+			}
+		}
+	}
+
+	return decode_nla_str(tcp, addr, len, NULL);
+}
+
+static
+DECL_NLA(ctrl_attr_op_id)
+{
+	const unsigned int *idx = opaque_data;
+	struct decode_nla_xlat_opts opts = {
+		.xlat = family_names[*idx].cmd[!entering(tcp)],
+		.dflt = "CTRL_CMD_???",
```

Comment:
It would be nice if we could define family-specific .dflt value in the `family_names` array.

```
static const struct {
	const char *family;
	const struct xlat *cmd[2];
        const char *dflt;
} family_names[] = {
...
	{ THERMAL_GENL_FAMILY_NAME, ARG_PAIR(genl_thermal_cmd), "THERMAL_GENL_CMD_???" },
...
}
...

	struct decode_nla_xlat_opts opts = {
		.xlat = family_names[*idx].cmd[!entering(tcp)],
		.dflt = family_names[*idx].dflt,
        ...
```